adr/0025-composition.md

```diff
@@ -0,0 +1,282 @@
+        newTag: latest
+      - name: mattermost/mattermost-enterprise-edition
+        newTag: 9.9.0
+    actions:
+      onDeploy:
+        after:
+          - description: Validate Mattermost Package
+            maxTotalSeconds: 300
+            wait:
+              cluster:
+                kind: Packages
+                name: mattermost
+                namespace: mattermost
+                condition: "'{.status.phase}'=Ready"
+
+  - name: mattermost-registry1
+
+    ## Previously a `required` component with the "registry1" `flavor`,
+    ## `mattermost-registry1` is now defined simply as a component that directly
+    ## extends the default `mattermost` component above.
+    #
+    # required: true
+    # import:
+    #   path: common
+    # only:
+    #   flavor: registry1
+    #   cluster:
+    #     architecture: amd64
+    extends: mattermost
+    required: false
+
+    ## In this case (and as is the case with most `registry1` component flavors)
+    ## the Helm chart values were only used to override image tags. We can now
+    ## take advantage of a more robust and declarative `images` replacements API.
+    #
+    # charts:
+    #   - name: mattermost-enterprise-edition
+    #     valuesFiles:
+    #       - values/registry1-values.yaml
+    #       ## sample contents of `values/registry1-values.yaml`
+    #       # mattermostApp:
+    #       #   image:
+    #       #     repository: registry1.dso.mil/ironbank/opensource/mattermost/mattermost
+    #       #     tag: 9.9.0
+    #       # initContainerImage:
+    #       #   repository: registry1.dso.mil/ironbank/redhat/ubi/ubi9-minimal
+    #       #   tag: 9.4
+    # images:
+    #   - registry1.dso.mil/ironbank/redhat/ubi/ubi9-minimal:9.4
+    #   - registry1.dso.mil/ironbank/opensource/mattermost/mattermost:9.9.0
+    images:
+      - name: appropriate/curl
+        newName: registry1.dso.mil/ironbank/redhat/ubi/ubi9-minimal
+        newTag: 9.4
+      - name: mattermost/mattermost-enterprise-edition
+        newName: registry1.dso.mil/ironbank/opensource/mattermost/mattermost
+        newTag: 9.9.0
+
+  ## Finally, the most interesting example of *optionally* enabling injection of
+  ## Mattermost plugins. This component is now "stateless" because it does nothing
+  ## by default and it is "self-contained" because it both builds the required image
+  ## and includes the necessary Helm chart values overrides
+  ## (i.e. `mattermostApp.extraInitContainers`).
+  - name: mattermost-plugins
+    required: false
+
+    ## Note that the semantics of `requires` is slightly different from `extends`.
+    ## The idea is that `extends` signals intent to _replace_ the original component
+    ## whereas `requires` signals that the originaly component(s) are being overlaid
+    ## with additional configuration.
+    requires: [ mattermost ]
+    charts:
+      - name: mattermost-enterprise-edition
+        valuesFiles:
+          - values/extra-plugins-values.yaml
```

Comment:
This part _could_ be interesting and is probably the main practical issue - optional components can't influence other components when they are selected - if a component could set a `chartOverride` map to be applied to later components that could be interesting - you would also need component name here as well to map those in correctly

Reply:
The [merge strategies](https://docs.zarf.dev/ref/components/#merge-strategies) for "named primitive arrays" (like `charts` in this example) is already well-defined. Why would further extending the API with something like `chartOverride` be preferable?

> you would also need component name here as well to map those in correctly

The component name that you are extending in this example is defined by `requires: [ mattermost ]`.